File: TaboolaSDKExampleObjective-C/TaboolaSDKExampleObjective-C/SDK Classic/ScrollView/ClassicConstraintsScrollViewController.m
//
//  ClassicConstraintsScrollViewController.m
//  TaboolaSDKExampleObjective-C
//
//  Copyright © 2020 Taboola. All rights reserved.
//

#import "ClassicConstraintsScrollViewController.h"
#import <TaboolaSDK/TaboolaSDK.h>
#import "RandomColor.h"

@interface ClassicConstraintsScrollViewController () <TBLClassicPageDelegate>

@property (weak, nonatomic) IBOutlet UIScrollView *scrollView;

@property (weak, nonatomic) IBOutlet UIView *top;
@property (weak, nonatomic) IBOutlet UIView *bottom;

// TBLClassicPage holds Taboola Widgets/Feed for a specific view
@property (nonatomic, strong) TBLClassicPage* classicPage;
// TBLClassicUnit object represnting Widget/Feed
@property (nonatomic, strong) TBLClassicUnit* taboolaWidgetPlacement;
@property (nonatomic, strong) TBLClassicUnit* taboolaFeedPlacement;

@end

@implementation ClassicConstraintsScrollViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self taboolaInit];
}

- (void)taboolaInit {
    _classicPage = [[TBLClassicPage alloc]initWithPageType:pageType pageUrl:pageUrl delegate:self scrollView:_scrollView];
    
    _taboolaWidgetPlacement = [_classicPage createUnitWithPlacementName:aboveArticePlacement mode:widgetMode_1x1];
    
    _taboolaFeedPlacement = [_classicPage createUnitWithPlacementName:placementFeedWithoutVideo mode:feedMode];
    _taboolaFeedPlacement.clipsToBounds = YES;

    [_taboolaWidgetPlacement fetchContent];
    [_taboolaFeedPlacement fetchContent];
    
    [self setTaboolaConstraintsToSuper];
}

- (void)setTaboolaConstraintsToSuper {
    [_top addSubview:_taboolaWidgetPlacement];
    _taboolaWidgetPlacement.translatesAutoresizingMaskIntoConstraints = NO;
    [_taboolaWidgetPlacement.topAnchor constraintEqualToAnchor:_top.topAnchor constant:0].active = YES;
    [_taboolaWidgetPlacement.bottomAnchor constraintEqualToAnchor:_top.bottomAnchor constant:0].active = YES;
    [_taboolaWidgetPlacement.leadingAnchor constraintEqualToAnchor:_top.leadingAnchor constant:0].active = YES;
    [_taboolaWidgetPlacement.trailingAnchor constraintEqualToAnchor:_top.trailingAnchor constant:0].active = YES;
    
    [_bottom addSubview:_taboolaFeedPlacement];
    _taboolaFeedPlacement.translatesAutoresizingMaskIntoConstraints = NO;
    [_taboolaFeedPlacement.topAnchor constraintEqualToAnchor:_bottom.topAnchor constant:0].active = YES;
    [_taboolaFeedPlacement.bottomAnchor constraintEqualToAnchor:_bottom.bottomAnchor constant:0].active = YES;
    [_taboolaFeedPlacement.leadingAnchor constraintEqualToAnchor:_bottom.leadingAnchor constant:0].active = YES;
    [_taboolaFeedPlacement.trailingAnchor constraintEqualToAnchor:_bottom.trailingAnchor constant:0].active = YES;
}

- (void)dealloc {
    NSLog(@"Dealloc");
    [self.classicPage reset];
}

#pragma mark - TBLClassicPageDelegate

- (void)classicUnit:(UIView *)classicUnit didLoadOrResizePlacementName:(NSString *)placementName height:(CGFloat)height placementType:(PlacementType)placementType{
    NSLog(@"%@", placementName);
}

- (void)classicUnit:(UIView *)classicUnit didFailToLoadPlacementName:(NSString *)placementName errorMessage:(NSString *)error{
    NSLog(@"%@", error);
}

- (BOOL)classicUnit:(UIView *)classicUnit didClickPlacementName:(NSString *)placementName itemId:(NSString *)itemId clickUrl:(NSString *)clickUrl isOrganic:(BOOL)organic customData: (NSDictionary *)customData {
    return YES;
}

@end

File: TaboolaSDKExampleSwift/TaboolaSDKExampleSwift/SDK Classic/ScrollView/ClassicConstraintsScrollViewController.swift
//
//  ClassicConstraintsScrollViewController.swift
//  TaboolaSDKExampleV4
//
//  Copyright © 2020 Taboola. All rights reserved.
//

import UIKit
import TaboolaSDK

class ClassicConstraintsScrollViewController: UIViewController {
    
    @IBOutlet weak var scrollView: UIScrollView!
    
    @IBOutlet weak var topWidget: UIView!
    @IBOutlet weak var bottomFeed: UIView!

    // TBLClassicPage holds Taboola Widgets/Feed for a specific view
    var classicPage: TBLClassicPage?
    
    // TBLClassicUnit object represnting Widget/Feed
    var taboolaWidgetPlacement: TBLClassicUnit?
    var taboolaFeedPlacement: TBLClassicUnit?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        taboolaInit()
    }
    
    func taboolaInit() {
        classicPage = TBLClassicPage.init(pageType: Constants.pageTypeArticle, pageUrl: Constants.pageUrl, delegate: self, scrollView: scrollView)
        
        taboolaWidgetPlacement = classicPage?.createUnit(withPlacementName: Constants.placementAboveArticle, mode: Constants.widgetMode_1x1)
        
        taboolaFeedPlacement = classicPage?.createUnit(withPlacementName: Constants.placementFeedWithoutVideo, mode: Constants.thumbsFeedMode)

        if let taboolaWidgetPlacement = taboolaWidgetPlacement {
            taboolaWidgetPlacement.fetchContent()
        }
        
        if let taboolaFeedPlacement = taboolaFeedPlacement {
            taboolaFeedPlacement.clipsToBounds = true
            taboolaFeedPlacement.fetchContent()
        }
        
        setTaboolaConstraintsToSuper()
    }
    
    func setTaboolaConstraintsToSuper() {
        if let taboolaWidgetPlacement = taboolaWidgetPlacement {
            topWidget.addSubview(taboolaWidgetPlacement)
            taboolaWidgetPlacement.translatesAutoresizingMaskIntoConstraints = false
            taboolaWidgetPlacement.topAnchor.constraint(equalTo: topWidget.topAnchor, constant: 0).isActive = true
            taboolaWidgetPlacement.bottomAnchor.constraint(equalTo: topWidget.bottomAnchor, constant: 0).isActive = true
            taboolaWidgetPlacement.leadingAnchor.constraint(equalTo: topWidget.leadingAnchor, constant: 0).isActive = true
            taboolaWidgetPlacement.trailingAnchor.constraint(equalTo: topWidget.trailingAnchor, constant: 0).isActive = true
        }
        
        if let taboolaFeedPlacement = taboolaFeedPlacement {
            bottomFeed.addSubview(taboolaFeedPlacement)
            taboolaFeedPlacement.translatesAutoresizingMaskIntoConstraints = false
            taboolaFeedPlacement.topAnchor.constraint(equalTo: bottomFeed.topAnchor, constant: 0).isActive = true
            taboolaFeedPlacement.bottomAnchor.constraint(equalTo: bottomFeed.bottomAnchor, constant: 0).isActive = true
            taboolaFeedPlacement.leadingAnchor.constraint(equalTo: bottomFeed.leadingAnchor, constant: 0).isActive = true
            taboolaFeedPlacement.trailingAnchor.constraint(equalTo: bottomFeed.trailingAnchor, constant: 0).isActive = true
        }
    }
}

extension ClassicConstraintsScrollViewController: TBLClassicPageDelegate {
    func classicUnit(_ classicUnit: UIView!, didLoadOrResizePlacementName placementName: String!, height: CGFloat, placementType: PlacementType) {
        print("Placement name: \(String(describing: placementName)) has been loaded with height: \(height)")
    }
    
    func classicUnit(_ classicUnit: UIView!, didFailToLoadPlacementName placementName: String!, errorMessage error: String!) {
        print(error as Any)
    }
    
    func classicUnit(_ classicUnit: UIView!, didClickPlacementName placementName: String!, itemId: String!, clickUrl: String!, isOrganic organic: Bool, customData: [AnyHashable : Any]!) -> Bool {
        return true
    }
}
